contracts/contracts/system_contracts/kaiabridge/Judge.sol

```diff
@@ -0,0 +1,340 @@
+// Copyright 2024 The klaytn Authors
+// This file is part of the klaytn library.
+//
+// The klaytn library is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Lesser General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// The klaytn library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with the klaytn library. If not, see <http://www.gnu.org/licenses/>.
+
+// SPDX-License-Identifier: LGPL-3.0-only
+pragma solidity 0.8.24;
+
+import "@openzeppelin/contracts-upgradeable/proxy/utils/Initializable.sol";
+import "./ReentrancyGuardUpgradeable.sol";
+import "@openzeppelin/contracts/utils/introspection/IERC165.sol";
+import "@openzeppelin/contracts-upgradeable/proxy/utils/UUPSUpgradeable.sol";
+import "@openzeppelin/contracts-upgradeable/access/OwnableUpgradeable.sol";
+import "./IGuardian.sol";
+import "./IJudge.sol";
+
+contract Judge is Initializable, ReentrancyGuardUpgradeable, UUPSUpgradeable, OwnableUpgradeable, IERC165, IJudge {
+    /// @custom:oz-upgrades-unsafe-allow constructor
+    constructor() { _disableInitializers(); }
+
+    function initialize(
+        address[] calldata initJudges,
+        address initGuardian,
+        uint8 _minJudgeRequiredConfirm
+    )
+        public
+        initializer
+    {
+        __ReentrancyGuard_init();
+        require(IERC165(initGuardian).supportsInterface(type(IGuardian).interfaceId), "PDT::Judge: Guardian contract address does not implement IGuardian");
+
+        for (uint8 i=0; i<initJudges.length; i++) {
+            judges.push(initJudges[i]);
+            isJudge[initJudges[i]] = true;
+        }
+        minJudgeRequiredConfirm = _minJudgeRequiredConfirm;
+        guardian = initGuardian;
+
+        // Fill the first index(0) with dummy tx
+        addTransaction(address(0), "", 0);
+
+        __Ownable_init();
+        __UUPSUpgradeable_init();
+    }
+
+    function _authorizeUpgrade(address newImplementation) internal virtual override onlyOwner {}
```

Comment:
Should this be `onlyGuardian`? Or is `owner` a separate entity who has the highest trust like guardian?

Reply:
Very good point. Currently, the owner only has the control to upgrade a contract. 
There are two options:
1. Use the EoA account as an owner which is controlled by multisig (not a multisig scheme in the bridge contract)
2. Use guardian authorization. However this is somewhat challenge because the contract calls this `__authorizeUpgrade()`, but the client side just only calls `upgrades.upgradeProxy()`.

Reply:
The upgradeable permission is so powerful that it can practically disable all security feature. Thus, if you haven't discussed the existence of an entity which has the utmost trust higher than guardian, and haven't discussed how it should be managed, then I think guardian should have the upgradeable feature.

Guardian contract can directly invoke [Bridge.upgradeToAndCall(newImpl)](https://docs.openzeppelin.com/contracts/5.x/api/proxy#UUPSUpgradeable-upgradeToAndCall-address-bytes-), without using OZ's `upgrades` library. If you're worried about missing the validations of `upgrades.upgradeProxy()` (storage validation, etc.), I think you can [manually validate](https://docs.openzeppelin.com/upgrades-plugins/1.x/api-hardhat-upgrades#validate-implementation).

In addition, I just noticed that Guardian also has an owner. I believe Guardian's ` _authorizeUpgrade()` should be `onlyGuardian` as well.